src/core/network/webhook.ts

```diff
@@ -5,46 +5,39 @@ const debug = d('telegraf:webhook')
 
 export default function (
   hookPath: string,
-  updateHandler: (update: Update, res: http.ServerResponse) => Promise<void>,
-  errorHandler: (err: SyntaxError) => void
+  updateHandler: (update: Update, res: http.ServerResponse) => Promise<void>
 ) {
-  return (
+  return async (
     req: http.IncomingMessage,
     res: http.ServerResponse,
-    next?: () => void
-  ): void => {
-    debug('Incoming request', req.method, req.url)
-    if (req.method !== 'POST' || req.url !== hookPath) {
-      if (typeof next === 'function') {
-        return next()
-      }
+    next = (): void => {
       res.statusCode = 403
       return res.end()
     }
+  ): Promise<void> => {
+    debug('Incoming request', req.method, req.url)
+    if (req.method !== 'POST' || req.url !== hookPath) {
+      return next()
+    }
     let body = ''
-    req.on('data', (chunk: string) => {
-      body += chunk.toString()
-    })
-    req.on('end', () => {
-      let update: Update
-      try {
-        update = JSON.parse(body)
-      } catch (error) {
-        res.writeHead(415)
-        res.end()
-        return errorHandler(error)
-      }
-      updateHandler(update, res)
-        .then(() => {
-          if (!res.finished) {
-            res.end()
-          }
-        })
-        .catch((err: unknown) => {
-          debug('Webhook error', err)
-          res.writeHead(500)
-          res.end()
-        })
-    })
+    for await (const chunk of req) {
+      body += String(chunk)
+    }
+    let update: Update
+    try {
+      update = JSON.parse(body)
+    } catch (error: unknown) {
+      res.writeHead(415)
+      res.end()
+      debug('Failed to parse request body:', error)
+      return
+    }
+    try {
+      await updateHandler(update, res)
+    } catch (err: unknown) {
+      debug('Webhook error', err)
+      res.writeHead(500)
+      res.end()
```

Comment:
And another one, this error should probably be rethrown:

```suggestion
      res.writeHead(500)
      res.end()
      throw err
```

Reply:
This makes sense to me. @smaznet what do you think?